core/controllers/reader_test.py

```diff
@@ -1104,58 +1104,98 @@ def test_get_recommendation_ids_with_invalid_author_recommended_ids(self):
-                }, expected_status_int=404
+                    'current_node_id': self.NODE_ID_2
+                }, csrf_token=csrf_token, expected_status_int=404
             )
+        self.logout()
 
-    def test_get_fails_when_node_does_not_exist_in_story_mode(self):
+    def test_put_fails_when_node_does_not_exist_in_story_mode(self):
+        self.login(self.NEW_USER_EMAIL)
+        csrf_token = self.get_new_csrf_token()
         with self.swap(constants, 'ENABLE_NEW_STRUCTURE_VIEWER_UPDATES', True):
-            self.get_json(
-                '/explorehandler/recommendations/%s' % self.EXP_ID_0, params={
+            self.put_json(
+                '/explorehandler/recommendations/%s' % self.EXP_ID_0, {
                     'story_id': self.STORY_ID,
-                    'current_node_id': 'invalid_node',
-                    'include_system_recommendations': False
-                }, expected_status_int=404
+                    'current_node_id': 'invalid_node'
+                }, csrf_token=csrf_token, expected_status_int=404
             )
+        self.logout()
+
 
-    def test_get_fails_when_story_is_not_published_in_story_mode(self):
+    def test_put_fails_when_story_is_not_published_in_story_mode(self):
         topic_services.unpublish_story(
             self.TOPIC_ID, self.STORY_ID, self.admin_id)
+        self.login(self.NEW_USER_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+        with self.swap(constants, 'ENABLE_NEW_STRUCTURE_VIEWER_UPDATES', True):
+            self.put_json(
+                '/explorehandler/recommendations/%s' % self.EXP_ID_0, {
+                    'story_id': self.STORY_ID,
+                    'current_node_id': self.NODE_ID_2
+                }, csrf_token=csrf_token, expected_status_int=404
+            )
+        self.logout()
 
+    def test_put_returns_empty_string_when_user_completes_story(self):
+        self.login(self.NEW_USER_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+        story_services.record_completed_node_in_story_context(
+            self.new_user_id, self.STORY_ID, self.NODE_ID_2)
+        story_services.record_completed_node_in_story_context(
+            self.new_user_id, self.STORY_ID, self.NODE_ID_1)
         with self.swap(constants, 'ENABLE_NEW_STRUCTURE_VIEWER_UPDATES', True):
-            self.get_json(
-                '/explorehandler/recommendations/%s' % self.EXP_ID_0, params={
+            json_response = self.put_json(
+                '/explorehandler/recommendations/%s' % self.EXP_ID_0, {
                     'story_id': self.STORY_ID,
-                    'current_node_id': self.NODE_ID_2,
-                    'include_system_recommendations': False
-                }, expected_status_int=404
+                    'current_node_id': self.NODE_ID_3
+                }, csrf_token=csrf_token
             )
+        self.assertEqual(len(json_response['summaries']), 0)
+        self.logout()
```

Comment:
Add a test for the case where the user completes a chapter twice. (I.e. they complete it the first time, then they play some other chapters, and then they go back and do a second playthrough.)

What recommendation are you planning to show in that case? My guess is none -- i.e. it's just "back to story", since:

(a) the next chapter won't make sense, since they've done it already
(b) the latest unlocked chapter in the story would make more sense for them to start from the story page instead, since they can remind themselves of what happened just before it.

Does that sound sensible to you, though?

Reply:
Currently, it shows (b)  in the recommendations, but yeah, if that chapter is far off, it might be  better to start from the story viewer page. Will remove that suggestion.

Reply:
Done, returned empty list when an already completed chapter is played.